kernel/src/log_segment.rs

```diff
@@ -460,28 +465,99 @@ impl LogSegment {
+                let (protocol, metadata) = try_protocol_metadata_from_crc(path.clone(), engine)?;
+                return Ok((Some(metadata), Some(protocol)));
+            }
+        }
+
+        // 2. construct a "protocol-metadata log segment" as either:
+        //    a) if we have a CRC file > last checkpoint version, we only need commits from
+        //       (CRC_version, target_version].
+        //    b) if we don't have a CRC, or have a CRC older than most recent checkpoint, we proceed
+        //       as normal (full log segment).
+        let protocol_metadata_log_segment: Cow<'_, Self> = match &self.latest_crc_file {
+            Some(latest_crc_file)
+                if latest_crc_file.version >= self.checkpoint_version.unwrap_or(0) =>
+            {
+                let mut log_segment = self.clone();
+                log_segment.checkpoint_version = None;
+                log_segment.checkpoint_parts.clear();
+                // only keep commits from CRC_version + 1 to target version
+                log_segment
+                    .ascending_commit_files
+                    .retain(|commit| commit.version > latest_crc_file.version);
+                // only keep compactions with end version > crc version
+                log_segment.ascending_compaction_files.retain(|compaction| {
+                    if let LogPathFileType::CompactedCommit { hi } = compaction.file_type {
+                        hi > latest_crc_file.version
+                    } else {
+                        false
+                    }
+                });
+                Cow::Owned(log_segment)
+            }
+            _ => Cow::Borrowed(self),
+        };
+
+        // 3. do log replay
+        let actions_batches = protocol_metadata_log_segment.replay_for_metadata(engine)?;
         let (mut metadata_opt, mut protocol_opt) = (None, None);
         for actions_batch in actions_batches {
             let actions = actions_batch?.actions;
             if metadata_opt.is_none() {
-                metadata_opt = Metadata::try_new_from_data(actions.as_ref())?;
+                if let Some(metadata) = Metadata::try_new_from_data(actions.as_ref())? {
+                    metadata_opt = Some(metadata);
+                }
             }
             if protocol_opt.is_none() {
                 protocol_opt = Protocol::try_new_from_data(actions.as_ref())?;
             }
             if metadata_opt.is_some() && protocol_opt.is_some() {
                 // we've found both, we can stop
-                break;
+                return Ok((metadata_opt, protocol_opt));
             }
         }
-        Ok((metadata_opt, protocol_opt))
+
+        // 4. if we didn't find one of Protocol/Metadata, we can read it from CRC
+        if let Some(latest_crc_file) = &self.latest_crc_file {
+            let path = &latest_crc_file.location;
+            let (protocol, metadata) = try_protocol_metadata_from_crc(path.clone(), engine)?;
```

Comment:
I'd recommend a note just to clarify.  I nearly got tripped up and proposed an incorrect code cleanup to just return protocol + metadata from crc lol

```suggestion
            // Note: if metadata_opt or protocol_opt is present, it must be from a newer Delta version 
            // than the CRC file. Thus metadata_op and protocol_opt always take precedence.
            let (protocol, metadata) = try_protocol_metadata_from_crc(path.clone(), engine)?;
```